src/uct/ib/dc/dc_mlx5_ep.c

```diff
@@ -906,6 +907,16 @@ ucs_status_t uct_dc_mlx5_ep_fc_pure_grant_send(uct_dc_mlx5_ep_t *ep,
     UCT_DC_MLX5_CHECK_DCI_RES(iface, ep);
     UCT_DC_MLX5_IFACE_TXQP_GET(iface, ep, txqp, txwq);
 
+    if (txwq->flags & UCT_IB_MLX5_TXWQ_FLAG_FAILED) {
+        /* If the selected DCI has FAILED flag set in TXWQ, it means that this
+         * DCI has already been assigned to the FC endpoint and it has not
+         * been reset yet, It is currently waiting for all operation compeltions
+         * as a part of handling failure */
+        ucs_assertv(dci != UCT_DC_MLX5_EP_NO_DCI, "fc_ep %p: dci %u", ep, dci);
+        uct_dc_mlx5_ep_pending_common(iface, ep, &fc_req->super.super, 0, 1);
```

Comment:
what is this functions is already invoked from pending context?
(from `uct_dc_mlx5_iface_fc_grant`)

Reply:
> what is this functions is already invoked from pending context?
> (from `uct_dc_mlx5_iface_fc_grant`)

when error is detected, group of FC EP is descheduled form the arbiter

Reply:
it is not obvious from this flow, i'd at least add some assert that it is not happening

Reply:
> it is not obvious from this flow, i'd at least add some assert that it is not happening

added assertion that group isn't scheduled

Reply:
why not just return NO_RESOURCE?
it should be same flow as if DCI WQ is full - not accepting new post sends (till it's reset)

Reply:
@dmitrygx WDYT to return NO_RESOURCE here instead of add_pending+return UCS_OK? so called will add to pending

Reply:
good catch, done - we could just return UCS_ERR_NO_RESOURCE.
adding to pending queue is already there